extensions/markdown-language-features/src/languageFeatures/copyFiles/shared.ts

```diff
@@ -61,30 +63,87 @@ export const mediaMimes = new Set([
 	'audio/x-wav',
 ]);
 
-export async function getMarkdownLink(document: vscode.TextDocument, ranges: readonly vscode.Range[], urlList: string, token: vscode.CancellationToken): Promise<{ additionalEdits: vscode.WorkspaceEdit; label: string } | undefined> {
+const smartPasteRegexes = [
+	{ regex: /\[.*\]\(.*\)/g, is_markdown_link: true }, // Is a Markdown Link
+	{ regex: /!\[.*\]\(.*\)/g, is_markdown_link: true }, // Is a Markdown Image Link
+	{ regex: /\[([^\]]*)\]\(([^)]*)\)/g, is_markdown_link: false }, // In a Markdown link
+	{ regex: /^```[\s\S]*?```$/gm, is_markdown_link: false }, // In a fenced code block
+	{ regex: /^\$\$[\s\S]*?\$\$$/gm, is_markdown_link: false }, // In a fenced math block
+	{ regex: /`[^`]*`/g, is_markdown_link: false }, // In inline code
+	{ regex: /\$[^$]*\$/g, is_markdown_link: false }, // In inline math
+];
+
+export interface SkinnyTextDocument extends vscode.TextDocument {
+
+	/**
+		 * Converts the position to a zero-based offset.
+		 *
+		 * The position will be {@link TextDocument.validatePosition adjusted}.
+		 *
+		 * @param position A position.
+		 * @return A valid zero-based offset.
+		 */
+	offsetAt(position: vscode.Position): number;
+
+	/**
+		 * Get the text of this document. A substring can be retrieved by providing
+		 * a range. The range will be {@link TextDocument.validateRange adjusted}.
+		 *
+		 * @param range Include only the text included by the range.
+		 * @return The text inside the provided range or the entire text.
+		 */
+	getText(range?: vscode.Range): string;
+
+	/**
+		 * The associated uri for this document.
+		 *
+		 * *Note* that most documents use the `file`-scheme, which means they are files on disk. However, **not** all documents are
+		 * saved on disk and therefore the `scheme` must be checked before trying to access the underlying file or siblings on disk.
+		 *
+		 * @see {@link FileSystemProvider}
+		 * @see {@link TextDocumentContentProvider}
+		 */
+	readonly uri: vscode.Uri;
+}
+
+export interface SmartPaste {
+
+	/**
+	 * `true` if the link is being pasted within a code block, math block, or a markdown link.
+	 */
+	pasteAsMarkdownLink: boolean;
+
+	/**
+	 * `true` if the link is being pasted over a markdown link.
+	 */
+	updateTitle: boolean;
+
+}
+
+export async function uriToWorkspaceEdit(document: SkinnyTextDocument, ranges: readonly vscode.Range[], urlList: string, token: vscode.CancellationToken): Promise<{ additionalEdits: vscode.WorkspaceEdit; label: string } | undefined> {
```

Comment:
For naming, try to describe what the function is doing. In this case the important parts are:

- This function takes a document, set of edit locations, and a uri list
- It then returns an edit which creates links at the edit locations

So maybe something very explicit like: `createEditAddingLinksForUriList`